### PR TITLE
Visit more targets when validating attributes

### PR DESCRIPTION
**Pull request by rylev**
This begins to address #80048, allowing for additional validation of attributes. 

There are more refactorings that can be done, though I think they should be tackled in additional PRs:
* ICE when a builtin attribute is encountered that is not checked 
* Move some of the attr checking done `ast_validation` into `rustc_passes`
  * note that this requires a bit of additional refactoring, especially of extern items which currently parse attributes (and thus are a part of the AST) but do not possess attributes in their HIR representation.  
* Rename `Target` to `AttributeTarget`
* Refactor attribute validation completely to go through `Visitor::visit_attribute`. 
  * This would require at a minimum passing `Target` into this method which might be too big of a refactoring to be worth it. 
  * It's also likely not possible to do all the validation this way as some validation requires knowing what other attributes a target has. 

r? @davidtwco

### CODE REVIEW

**Comment by davidtwco** [comment]
In previous PRs which have checked more targets when validating attributes (#77015 and #79073), we did a crater run to check for regressions - this is technically a breaking change. Might be worth doing that here too.

**Comment by nagisa** [comment]
@craterbot queue

**Comment by nagisa** [comment]
@craterbot check

**Comment by ehuss** [comment]
Just wanted to give a heads up, I think PR #80641 is doing something almost identical, you may want to coordinate, and at least make sure the same tests are included and the issues (like "inline") are covered.

**Comment by rylev** [comment]
Here's [an issue](https://github.com/rust-lang/rust/issues/65833) on `#[inline]` on associated constants that we might want to check we fix with this.

**Comment by davidtwco** [comment]
Looks like this regresses [`olm`](https://crater-reports.s3.amazonaws.com/pr-80920/try%23bb9dbf1d07454a99066eabaabd5819851281494b/reg/olm-rs-1.0.0/log.txt) (three dependencies, now also failing) and [`RustWebServer`](https://crater-reports.s3.amazonaws.com/pr-80920/try%23bb9dbf1d07454a99066eabaabd5819851281494b/gh/nickyc975.RustWebServer/log.txt) (no dependencies).

**Comment by rylev** [comment]
The results of the crater run are pretty clear that this doesn't have a widespread impact. There were no crates that seemed to be explicitly effected. 

It seems 3 crates had `#[used]` annotations on struct fields which is now an error. There were no other errors (the other failures were transient).

**Comment by davidtwco** [comment]
I'll nominate this for the language team to decide if the breaking changes are acceptable.

**Comment by sanxiyn** [comment]
Note that `#[used]` provides access to `llvm.used`, but support in Rust is known to be incomplete. I think, at least, #68322 should be fixed.

**Comment by nikomatsakis** [comment]
cc @jhaye and @poljar -- the regression on olm-rs seems real and affects the most recent version. Are you able to issue a 1.0.1 release that removes the `#[used]` from being used on fields?

**Comment by Mark-Simulacrum** [comment]
We discussed this during lang meeting today and concluded that we can move forward (we want to ping authors of the affected crates, but I believe have done so in Niko's comment and @joshtriplett is going to file an issue on the other project).

We also noted that generally speaking this breakage is unfortunate and we may want to move to only breaking crates over edition boundaries in cases like this, but we're not prepared to reconsider the current pre-existing behavior here for inline in other contexts. It's unclear to me that we get a huge amount of benefit from avoiding the breakage in cases like this too.

**Comment by poljar** [comment]
> cc @jhaye and @poljar -- the regression on olm-rs seems real and affects the most recent version. Are you able to issue a 1.0.1 release that removes the `#[used]` from being used on fields?

Sure, PR is up. Sorry for the inconvenience.

**Comment by jhaye** [comment]
> > cc @jhaye and @poljar -- the regression on olm-rs seems real and affects the most recent version. Are you able to issue a 1.0.1 release that removes the `#[used]` from being used on fields?
> 
> Sure, PR is up. Sorry for the inconvenience.

v1.0.1 was just published.

**Comment by davidtwco** [comment]
I think we can probably land this then, could you rebase @rylev?

File: compiler/rustc_passes/src/check_attr.rs [hunk 1]
```diff
@@ -1101,17 +1101,6 @@ impl Visitor<'tcx> for CheckAttrVisitor<'tcx> {
         intravisit::walk_arm(self, arm);
     }
 
-    fn visit_macro_def(&mut self, macro_def: &'tcx hir::MacroDef<'tcx>) {
```
Comment (rylev) [inline compiler/rustc_passes/src/check_attr.rs:1104]: This was just moved to lower in the file during rebasing. I can move it back if desired.

**Comment by rylev** [comment]
@davidtwco I've rebased. The PR is a lot less exciting than it started as, but there might be a reason to still merge it. Perhaps though we should further refactor how attributes are validated in other parts of the compiler.

**Review by davidtwco — APPROVED** [review]

**Comment by eddyb** [comment]
Sorry I missed these parts of the description, but hopefully this helps for future PRs:

> This would require at a minimum passing `Target` into this method which might be too big of a refactoring to be worth it.

You can always keep data in the visitor itself, but that's harder to make sure nothing goes wrong with.

> It's also likely not possible to do all the validation this way as some validation requires knowing what other attributes a target has.

It's a bit weird IMO that `visit_attributes` doesn't exist. That might be a good way to do it.

**Pull request merged**